File: examples/python/RGBD/rgbd_pcl_processing.py
import depthai as dai
import numpy as np
from argparse import ArgumentParser


class CustomPCLProcessingNode(dai.node.ThreadedHostNode):
    def __init__(self):
        dai.node.ThreadedHostNode.__init__(self)
        self.inputPCL = self.createInput()
        self.outputPCL = self.createOutput()
        self.thresholdDistance = 3000.0

    def run(self):
        while self.mainLoop():
            try:
                inPointCloud = self.inputPCL.get()
            except dai.MessageQueue.QueueException:
                return # Pipeline closed
            if inPointCloud is not None:
                outPointCloud = dai.PointCloudData()
                points, colors = inPointCloud.getPointsRGB()

                mask = np.linalg.norm(points, axis=1) < self.thresholdDistance
                updatedPoints = points[mask]
                updatedColors = colors[mask]

                outPointCloud.setPointsRGB(updatedPoints, updatedColors)
                self.outputPCL.send(outPointCloud)

parser = ArgumentParser()
parser.add_argument("--webSocketPort", type=int, default=8765)
parser.add_argument("--httpPort", type=int, default=8082)
args = parser.parse_args()

with dai.Pipeline() as p:
    remoteConnector = dai.RemoteConnection(
        webSocketPort=args.webSocketPort, httpPort=args.httpPort
    )
    # The Depth node manages its own stereo cameras and backend internally, and
    # RGBD aligns its depth to the color camera internally.
    # Pick a COLOR-capable socket for the color camera so it does not take the
    # socket the Depth node needs (e.g. the ToF sensor on ToF-only devices, where
    # the default AUTO socket would otherwise grab the only ToF-capable sensor).
    colorSocket = dai.CameraBoardSocket.CAM_A
    for features in p.getDefaultDevice().getConnectedCameraFeatures():
        if dai.CameraSensorType.COLOR in features.supportedTypes:
            colorSocket = features.socket
            break
    color = p.create(dai.node.Camera).build(colorSocket)
    depth = p.create(dai.node.Depth)
    rgbd = p.create(dai.node.RGBD).build(color, depth)
    customNode = p.create(CustomPCLProcessingNode)

    # Link rgbd.pcl to the input of CustomPCLProcessingNode
    rgbd.pcl.link(customNode.inputPCL)

    remoteConnector.addTopic("pcl", rgbd.pcl, "common")
    remoteConnector.addTopic("processed_pcl", customNode.outputPCL, "common")

    p.start()
    remoteConnector.registerPipeline(p)

    while p.isRunning():
        key = remoteConnector.waitKey(1)
        if key == ord("q"):
            print("Got q key from the remote connection!")
            break

File: examples/cpp/RGBD/rgbd_pcl_processing.cpp

#include <csignal>
#include <depthai/remote_connection/RemoteConnection.hpp>
#include <iostream>

#include "depthai/depthai.hpp"

class CustomPCLProcessingNode : public dai::NodeCRTP<dai::node::ThreadedHostNode, CustomPCLProcessingNode> {
   public:
    constexpr static const char* NAME = "CustomPCLProcessingNode";
    constexpr static const float thresholdDistance = 3000.0f;

   public:
    Input inputPCL{*this, {"inPCL", DEFAULT_GROUP, DEFAULT_BLOCKING, DEFAULT_QUEUE_SIZE, {{{dai::DatatypeEnum::PointCloudData, true}}}}};
    Output outputPCL{*this, {"outPCL", DEFAULT_GROUP, {{{dai::DatatypeEnum::PointCloudData, true}}}}};

    void run() override {
        while(isRunning()) {
            auto pclIn = inputPCL.get<dai::PointCloudData>();
            auto pclOut = std::make_shared<dai::PointCloudData>();
            if(pclIn != nullptr) {
                const auto& pclData = pclIn->getPointsRGB();
                std::vector<dai::Point3fRGBA> updatedPoints;
                const float thresholdSquared = thresholdDistance * thresholdDistance;
                updatedPoints.reserve(pclData.size());
                for(const auto& point : pclData) {
                    const float distance = point.x * point.x + point.y * point.y + point.z * point.z;
                    if(distance <= thresholdSquared) {
                        updatedPoints.emplace_back(point.x, point.y, point.z, point.r, point.g, point.b, point.a);
                    }
                }
                updatedPoints.shrink_to_fit();
                pclOut->setPointsRGB(updatedPoints);
                outputPCL.send(pclOut);
            }
        }
    }
};

// Signal handling for clean shutdown
static bool isRunning = true;
void signalHandler(int signum) {
    isRunning = false;
}

int main() {
    using namespace std;
    // Default port values
    int webSocketPort = 8765;
    int httpPort = 8082;

    // Register signal handler
    std::signal(SIGINT, signalHandler);

    // Create RemoteConnection
    dai::RemoteConnection remoteConnector(dai::RemoteConnection::DEFAULT_ADDRESS, webSocketPort, true, httpPort);
    // Create pipeline
    dai::Pipeline pipeline;
    auto rgbd = pipeline.create<dai::node::RGBD>()->build(true, dai::node::StereoDepth::PresetMode::DEFAULT);
    auto customNode = pipeline.create<CustomPCLProcessingNode>();

    rgbd->pcl.link(customNode->inputPCL);

    remoteConnector.addTopic("pcl", rgbd->pcl);
    remoteConnector.addTopic("processed_pcl", customNode->outputPCL);
    pipeline.start();
    remoteConnector.registerPipeline(pipeline);
    auto device = pipeline.getDefaultDevice();
    device->setIrLaserDotProjectorIntensity(0.7);
    // Main loop
    while(isRunning && pipeline.isRunning()) {
        int key = remoteConnector.waitKey(1);
        if(key == 'q') {
            std::cout << "Got 'q' key from the remote connection!" << std::endl;
            break;
        }
    }

    std::cout << "Pipeline stopped." << std::endl;
    return 0;
}
